app/services/fact_importer.rb

```diff
@@ -37,11 +37,16 @@ def initialize(host, facts = {})
 
   # expect a facts hash
   def import!
-    delete_removed_facts
-    add_new_facts
-    update_facts
+    ActiveRecord::Base.transaction do
+      delete_removed_facts
+      update_facts
+      add_new_facts
+    end
 
-    raise ::Foreman::Exception.new(N_("Import of facts failed for host %s"), @host.name) if @error
+    if @error
+      Foreman::Logging.exception("Error during fact import for #{@host.name}", @error)
+      raise ::Foreman::WrappedException.new(@error, N_("Import of facts failed for host %s: #{@error}"), @host.name)
```

Comment:
Interpolation within an extracted string is incorrect (`#{@error}`), and unnecessary here as you're using WrappedException. Just restore the original, `N_("Import of facts failed for host %s")`.